webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/reports/execution/ReportsExecutionRepository.scala

```diff
@@ -109,47 +105,30 @@ class CachedReportsExecutionRepository(
    * can be used for a given node) is given by the presence of the
    * nodeid in map's keys.
    */
-  private var cache = Map[NodeId, Option[AgentRunWithNodeConfig]]()
+  private val cacheRef =
+    Ref.Synchronized.make(Map[NodeId, Option[AgentRunWithNodeConfig]]()).runNow
+
+  override def clearCache(): Unit = cacheRef.set(Map()).runNow
 
-  override def clearCache(): Unit = semaphore
-    .withPermit(IOResult.attempt {
-      cache = Map()
+  override def getNodesLastRun(nodeIds: Set[NodeId]): IOResult[Map[NodeId, Option[AgentRunWithNodeConfig]]] = cacheRef
+    .modifyZIO(f = cache => {
+      for {
+        n1          <- currentTimeMillis
+        _           <- ComplianceDebugLoggerPure.trace(s"cache last run ; ${cacheRef}")
+        _           <- ComplianceDebugLoggerPure.trace(s"node id diff last run ; ${nodeIds.diff(cache.keySet)}")
+        runs        <- readBackend.getNodesLastRun(nodeIds.diff(cache.keySet))
+        n2          <- currentTimeMillis
+        _           <- TimingDebugLoggerPure.trace(s"CachedReportsExecutionRepository: get nodes last run in: ${n2 - n1}ms")
+        cacheUpdated = cache ++ runs
+      } yield (cacheUpdated.view.filterKeys(x => nodeIds.contains(x)).toMap, cacheUpdated)
```

Comment:
`cacheUpdated.view.filterKeys(x => nodeIds.contains(x)).toMap` is (almost) the same as just `runs`. 
The only difference could be the "None" comming from the node ids excluded in `nodeIds.diff(cache.keySet)` which I think we don't care of. 
I would simplify showing the intent with just `yield (runs, cacheUpdated)`